### PR TITLE
[Semi Auto] Matmul & Embedding InferBackward Rule 

### PR DESCRIPTION
**Pull request by JZ-LIANG**
### PR types
Function optimization

### PR changes
Others

### Description
Pcard-70448
inferbackward rule for matmul & embedding

### CODE REVIEW

File: test/auto_parallel/spmd_rules/test_matmul_rule.py [hunk 1]
```diff
@@ -226,21 +227,158 @@ def test_matmul_infer_forward(self):
...
+        self.assertEqual(
+            infered_input_dist_attrs[1].dims_mapping, [-1, -1, -1, 1]
+        )
+        self.assertEqual(
+            infered_output_dist_attrs[0].dims_mapping, [-1, 0, -1, 1]
+        )
+
+        # trans_x = true, trans_y = true, abmn[-1, -1, 0, 1] --> abmk[-1, -1, -1, 0], a1kn[-1, -1, 1, -1]
+        self.out_dist_tensor_spec.shape = [512, 48, 64, 48]
+        self.x_dist_tensor_spec.shape = [512, 48, 64, 32]
```
Comment (zhiqiu) [inline test/auto_parallel/spmd_rules/test_matmul_rule.py]:
```suggestion
        self.x_dist_tensor_spec.shape = [512, 48, 32, 64]
        self.y_dist_tensor_spec.shape = [512, 1, 48, 32]
```

File: test/auto_parallel/spmd_rules/test_embedding_rule.py [hunk 2]
```diff
@@ -110,6 +113,66 @@ def test_embedding_infer_forward(self):
...
+
+        self.attrs = {
+            'padding_idx': -1,
+            'sparse': False,
+        }
+
+        # data parallel
+        self.out_dist_tensor_spec.set_dims_mapping([1, -1, -1])
+        result_dist_attrs = self.rule1.infer_backward(
+            [self.out_dist_tensor_spec], self.attrs
```
Comment (zhiqiu) [inline test/auto_parallel/spmd_rules/test_embedding_rule.py]: the inputs of infer_backward includes input_specs and output_specs

File: test/auto_parallel/spmd_rules/test_matmul_rule.py (hunk 1 above)
Comment (JZ-LIANG) [reply test/auto_parallel/spmd_rules/test_matmul_rule.py]: fixed

File: test/auto_parallel/spmd_rules/test_embedding_rule.py (hunk 2 above)
Comment (JZ-LIANG) [reply test/auto_parallel/spmd_rules/test_embedding_rule.py]: fixed

**Review by zhiqiu — DISMISSED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**